### DISCUSSION TITLE
the quality of the stroke is a bit too low

### DISCUSSION THREAD

**Discussion by Wladefant** [personalizedrefrigerator/js-draw#47] [Ideas & Feature Requests]
I can clearly see the pixels of the stroke, that gives it a very cheap look. 
For comparison of strokes of teh same size

 js draw:
![SmartSelect_20231026_152520_Joplin (Beta)](https://github.com/personalizedrefrigerator/js-draw/assets/79611093/0bf7a872-c066-42fd-89ff-043eb5086adf)

and samsung ntoes:
![SmartSelect_20231026_152530_Samsung Notes](https://github.com/personalizedrefrigerator/js-draw/assets/79611093/f0ef0402-7535-4b9f-886b-01d48e9d7c2b)

**Comment by personalizedrefrigerator** [comment]
There are a few things that could be causing this:

# 1. Not scaling the canvas area by the device-pixel ratio

If the stroke was *just* drawn and the canvas hasn't been moved (and you're on a high resolution display), `js-draw` probably isn't [correctly scaling the canvas](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Optimizing_canvas#scaling_for_high_resolution_displays).

If this is the case, I'm hopeful that this will be easy to fix. It may make `js-draw` slower though.

# 2. `js-draw`'s rendering cache

If you aren't on a high resolution display, and the canvas was moved after the stroke was first drawn, it could be `js-draw`'s internal cache. It does require a large-ish number of strokes to activate the cache (so this would only be the case if there are a large number of strokes just off-screen).

`js-draw` internally uses a [2-d tree](https://en.wikipedia.org/wiki/K-d_tree) to divide the drawing area into smaller canvases. This allows `js-draw` to avoid potentially expensive re-renders of parts of the screen with many strokes.

With the goal of avoiding re-renders, [the cache currently allows the tree's canvases to be scaled up by 1.3×](https://github.com/personalizedrefrigerator/js-draw/blob/539e3bac882f08ebd16a7035ec240cf6bc923953/packages/js-draw/src/rendering/Display.ts#L76), at some zoom levels, this leads to blurriness, and at other zoom levels (when the cache nodes aren't being scaled up), this is fine.

**Reply by Wladefant** [reply]
I am on a high resolution screen (Samsung Galaxy tab s8+). Not many strokes

**Reply by personalizedrefrigerator** [reply]
Incorrect canvas scaling should be addressed by [this commit](https://github.com/personalizedrefrigerator/js-draw/commit/676beb91bab39b7eef91d019fd40c830f4d0b93c).

**Reply by personalizedrefrigerator** [reply]
It's in the latest prerelease: https://github.com/personalizedrefrigerator/joplin/releases/tag/prerelease-2

Note that this prerelease also fixes a rather significant sync issue. The issue should only affect users who previously had the [second-latest prerelase](https://github.com/personalizedrefrigerator/joplin/releases/tag/prerelease-1) installed, and only the images edited with that prerelease.

**Reply by Wladefant** [reply]
Yes sync issue, Just wanted to report it

**Comment by Wladefant** [comment]
The Quality is much better now, but when zooming out it gets a bit like it is too low quality. But bear in mind please, if something affects latency dont add it, because latency is priority number one and it still definitely lacks behind samsungs latency though I am not sure it is possible to achieve this, as we are based on react native
![SmartSelect_20231031_102856_Joplin (Beta)](https://github.com/personalizedrefrigerator/js-draw/assets/79611093/1f8f0ec5-fa37-4500-b8bc-8bb833539d97)


https://github.com/personalizedrefrigerator/js-draw/assets/79611093/137535b9-e79c-447f-9777-38ffec034101

**Reply by personalizedrefrigerator** [reply]
**That** is probably because of the cache (see explanation above).

You can see whether the cache is being, or has been, used  by clicking on "About" under the "Page" menu, then clicking "developer information":

![screenshot: About button under "page" menu](https://github.com/personalizedrefrigerator/js-draw/assets/46334387/28ac79c5-07ff-499c-8352-eb5858ee4c28)
![screenshot: Developer info menu](https://github.com/personalizedrefrigerator/js-draw/assets/46334387/ce8129d0-870f-47b6-b6ee-7d70d4604572)

Above, the developer information menu states that 90/90 cache records have been used. This means that the cache has been used since the image has been opened.

I'm hopeful that this can be improved without degrading performance (perhaps by fixing cache-related performance issues elsewhere :) ).
